test-data/unit/check-dataclasses.test

```diff
@@ -744,6 +744,17 @@ s: str = a.bar()  # E: Incompatible types in assignment (expression has type "in
 
 [builtins fixtures/dataclasses.pyi]
 
+[case testDataclassGenericCovariant]
+from dataclasses import dataclass
+from typing import Generic, TypeVar
+
+T_co = TypeVar("T_co", covariant=True)
+
+@dataclass
+class MyDataclass(Generic[T_co]):
+    a: T_co
```

Comment:
I'm struggling to find an example where mypy's current behaviour actually causes some unsafe behaviour, but it feels weird to me that mypy will reject this:

```py
from typing import TypeVar, Protocol

T_co = TypeVar("T_co", covariant=True)

class Foo(Protocol[T_co]):  # error: Covariant type variable "T_co" used in protocol where invariant one is expected
    x: T_co
```

But is fine with this:

```py
from dataclasses import dataclass
from typing import TypeVar, Generic

T_co = TypeVar("T_co", covariant=True)

@dataclass
class Bar(Generic[T_co]):
    x: T_co
```

It feels like the argument that the former is unsafe should probably also apply to the latter? But as I say, I'm struggling to come up with a real example where mypy is currently not flagging unsafe behaviour with generic dataclasses.

Reply:
Both are unsafe, see also https://github.com/python/mypy/issues/3208. But for regular (i.e. nominal) classes it is hard to prohibit this, since it would force people to use `@property` or `Final` explicitly everywhere if the attribute is effectively read-only. Protocols were a new thing, so we were able to implement them safely from the start (also an additional argument is that bugs with structural subtyping are usually trickier to find, so our handling of structural subtyping should be stricter that our handling of nominal subtyping).

Btw I was thinking about actually trying to prohibit covariant mutable overrides in `--strict` mode first, and see what will be the fallout and how people will react to this.

Reply:
Btw I'm merely testing here a regression that surfaced in mypy_primer.

Reply:
> Btw I'm merely testing here a regression that surfaced in mypy_primer.

Yes, I was questioning whether it _was_ actually a regression, or whether the new behaviour that we had with your PR was actually a feature rather than a bug :-)

Reply:
Even if we went to report it, probably not like _this_.

What was being reported were covariant parameters of the synthetic `__mypy-replace` and `__mypy-post_init` methods (which are added in a less hacky manner now and probably that's why they hit this check).